docs/specifications/tests/Delegation-TP/delegation01.md

```diff
@@ -1,49 +1,118 @@
-## DELEGATION01: Minimum number of name servers   
+# DELEGATION01: Minimum number of name servers   
 
-### Test case identifier
+## Test case identifier
 
-**DELEGATION01:** Minimum number of name servers
+**DELEGATION01**
 
-### Objective
+## Objective
 
-Section 4.1 of [RFC 1034](https://tools.ietf.org/html/rfc1034) specifies that
-there must be at least minimum two name servers for a domain. This test is
-done to verify this condition
+Section 4.1 of [RFC 1034] specifies that there must be a minimum of two name servers 
+for a domain. This test is done to verify this condition.
 
-### Inputs
+The RFC ([RFC 1034]) predates IPv6. Since IPv4 and IPv6 work as separate networks, this
+test case has been extended to test for two name servers that resolve into IPv4 addresses
+and IPv6 addresses respectively.
 
-The domain name to be tested.
 
-### Ordered description of steps to be taken to execute the test case
+## Inputs
 
-1. Obtain the complete set of name servers from parent using
-   [Method2](../Methods.md).
-2. If the number of name servers (distinct NS records) returned in previous step is less than two, the
-   record as an error.
-3. Obtain the complete set of name servers from the child 
-   using [Method3](../Methods.md).
-4. If the number of name servers (distinct NS records) returned in the previous step is less than two, the
-   record as an error.
-5. If an error was recorded in step 2 or step 4, then the test case fails.
- 
-### Outcome(s)
+The domain name to be tested ("Child Zone").
 
-If the total amount of name servers used for the domain are two or more,
-then the test case succeeds.
+## Ordered description of steps to be taken to execute the test case
 
-### Special procedural requirements
+ 1. Using [Method2], obtain the complete set of names of the name servers 
+    from the delegation of the *Child Zone*.
+
+ 2. Count the name server names:
+    1. If zero or one, emit *NOT_ENOUGH_NS_DEL*.
+    2. If two or more, emit *ENOUGH_NS_DEL*.
+
+ 3. Using [Method4], obtain the IP addresses for the name servers of the 
+    delegation of the *Child Zone*.
+
+ 4. Count the number of name server names that resolve into at least one IPv4 
+    address:
+    1. If zero, emit *NO_IPV4_NS_DEL*.
+    2. If one, emit *NOT_ENOUGH_IPV4_NS_DEL*.
+    3. If two or more, emit *ENOUGH_IPV4_NS_DEL*.
+
+ 5. Count the number of name server names that resolve into at least one IPv6 
+    address:
+    1. If zero, emit *NO_IPV6_NS_DEL*.
+    2. If one, emit *NOT_ENOUGH_IPV6_NS_DEL*.
+    3. If two or more, emit *ENOUGH_IPV6_NS_DEL*.
+
+ 6. Using [Method3], obtain the complete set of names of the name servers
+    from the *Child Zone* for the *Child Zone*. 
+
+ 7. Count the name server names:
+    1. If zero or one, emit *NOT_ENOUGH_NS_CHILD*.
+    2. If two or more, emit *ENOUGH_NS_CHILD*.
+
+ 8. Using [Method5], obtain the IP addresses for the name servers from 
+    the *Child Zone* for the *Child Zone*.
```

Comment:
I interpret this step to obtains a set of IP-addresses. That's insufficient information to make the determination in the next step. Maybe we need a new method that returns a set of (name, address) pairs?

Reply:
Maybe we should have just two methods, one method that returns the name server names and its IP addresses from the delegation point of view and one method that returns the name server names and its IP addresses from the child zone point of view. The test cases could choose to call the methods in one of three ways:
* Set of IP addresses.
* Set of name servers (names).
* Set of name servers with IP adresses.

Reply:
I agree that there's a symmetry here, but I don't think the right way to express that symmetry is to cram three basic methods into one super-method and having two of those super-methods. The simple solution is just having six methods and expressing the symmetry in their naming.

The structure of the return types is also interesting. We seem to agree that the first two methods return sets of IP addresses and sets of names, respectively. For the third kind you seem to think of the return type as a set of (name, set of IP addresses)-pairs where all names are unique. A mathematically equivalent (and much simpler) type is a set of (name, IP address)-pairs without the uniqueness requirement. This also happens to match Engine's internal representation. A `Zonemaster::Engine::Nameserver` represents a (name, IP address)-pair together with a bunch of additional properties. I suggest that we use the simpler type in the specifications as well.

Lastly, we need names for things. The two points of view need to be named, as well as the three basic methods. We also need some stylistic convention for the names - like use of abbreviations, casing and word separators. I don't really have a suggestion for the styling, but here's my two cents for the naming and return types:
* Delegation name server addresses: `Set[IP address]`
* Delegation name server names: `Set[name]`
* Delegatoin name servers: `Set[(name, IP address)]`
* In-zone name server addresses: `Set[IP address]`
* In-zone name server names: `Set[name]`
* In-zone name servers: `Set[(name, IP address)]`